src/V3DynamicClass.cpp

```diff
@@ -0,0 +1,93 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+// DESCRIPTION: Verilator: Prepare AST for dynamic scheduling features
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2022 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+// V3DynamicClass's Transformations:
+//
+//      Each class with an event member:
+//          In the destructor, notify the event dispatcher that the event var doesn't exist
+//          anymore.
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Global.h"
+#include "V3DynamicClass.h"
+#include "V3Ast.h"
+
+//######################################################################
+// Manage the lifetime of class member events
+
+class DynamicClassEventVisitor final : public VNVisitor {
+private:
+    // STATE
+    AstClass* m_classp = nullptr;  // Current class
+    AstNode* m_resetStmtsp = nullptr;  // Statements that reset member events
+    AstCFunc* m_constructor = nullptr;  // Current class constructor
+    AstCFunc* m_destructor = nullptr;  // Current class destructor
+
+    // METHODS
+    VL_DEBUG_FUNC;  // Declare debug()
+
+    // VISITORS
+    virtual void visit(AstClass* nodep) override {
+        VL_RESTORER(m_classp);
+        VL_RESTORER(m_resetStmtsp);
+        VL_RESTORER(m_destructor);
```

Comment:
Can `AstClass` nest? If not I would prefer an assertion that these are `nullptr`, and reset back to `nullptr` at the end, as that makes it more explicit what input is expected.

Also `m_cosntructor` seems to be missing a restore?

Reply:
They can't nest currently, but that could change once Verilator supports nested classes, depending on the implementation.

Reply:
In C++ nested classes are a purely lexical construct (AFAIR) so we will very likely want to extract them from SV too.